mindmeld/components/nlp.py

```diff
@@ -1416,9 +1429,82 @@ def process_query(self, query, allowed_nlp_classes=None, dynamic_resource=None,
-            query[0], entities=processed_entities, confidence=confidence
-        )
+        return ProcessedQuery(query[0], entities=processed_entities, confidence=confidence)
+
+    def _find_entities_in_text(self, query, dynamic_resource,
+                               allowed_nlp_classes, max_ngram_search):
+        """
+        This function finds all entities in the query using rule-based matching based on the user
+        provided allowed_nlp_classes dict. There are two matching criterion:
+        1. gazzeteer based matching: We compare query ngrams to the gazetteers
+        2. duckling based matching: We compare query ngrams to the duckling candidates
+        Since this approach is rule-based, there are bound to be overlapping entity candidates.
+        So we find the largest non-overlapping set of candidates.
+
+        Args:
+            query (tuple of Query): The n-best queries
+            dynamic_resource (dict): gazetteer resources
+            allowed_nlp_classes (dict, optional): A dictionary of the NLP hierarchy that is \
+                selected for NLP analysis. An example: ``{'smart_home': {'close_door': {}}}`` \
+                where smart_home is the domain and close_door is the intent.
+            max_ngram_search (int): The max ngram value we want to break the input query
+
+        Returns:
+            list: A list of lists of non-overlapping entities for each n-best transcript
+        """
+        # This code block implements allowed entities described here:
+        # https://github.com/cisco/mindmeld/pull/280
+
+        # We extract the `entity` key from each token since that represents the normalized text of
+        # each token which we will compare against the normalized text of the query
+
+        def tokenize_to_tuples(text):
+            return tuple([token["entity"] for token in
+                          self.resource_loader.query_factory.tokenizer.tokenize(text)])
+
+        dynamic_gazetteer = dynamic_resource.get(GAZETTEER_RSC, {}) if dynamic_resource else {}
+        n_best_entities = [[] for _ in range(len(query))]
+
+        for entity in allowed_nlp_classes:
+            # check if entity is a system entity
+            if entity.startswith(SYSTEM_ENTITY_PREFIX):
+                for idx, n_best_query in enumerate(query):
+                    for sys_entity in n_best_query.system_entity_candidates:
+                        if sys_entity.entity.type == entity:
+                            n_best_entities[idx].append(sys_entity)
+                continue
+
+            # check if entity is in the gazetteers
+            # TODO: We are tokenizing every phrase in the gazetteer.
+            # This is a large CPU intensive op
+            consolidate_list = {**self.resource_loader.get_gazetteer(entity)['pop_dict'],
+                                **dynamic_gazetteer.get(entity, {})}
+            consolidated_set = {tokenize_to_tuples(key) for key in consolidate_list.keys()}
```

Comment:
Minor Nit: We can remove `.keys()`

`consolidated_set = {tokenize_to_tuples(key) for key in consolidate_list}`

Reply:
done